### ISSUE TITLE
Documentation of `mem::uninitialized` potentially confusing for ZSTs

### ISSUE THREAD

**Issue by steffahn** [rust-lang/rust#83995]
I recently wanted to create a value of a zero-sized type. (Making sure that the type is `'static` and `Copy` and my API already got passed an instance of it at some point earlier.) After some thinking, I’ve ended up using `MaybeUninit::uninit().assume_init()` to create a value of the type. But I don’t feel 100% satisfied with this choice. The reason is the following: The [`mem::unitialized`](https://doc.rust-lang.org/std/mem/fn.uninitialized.html) docs tell me that
* >it has the same effect as [`MaybeUninit::uninit().assume_init()`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.uninit)

  and
* >the function basically cannot be used correctly

Now from my use case, I would argue that for the case of zero-sized types, `mem::unitialized` _can_ indeed be used correctly.

Is this correct? If it is, then maybe adjusting the documentation would help avoid confusion.

@rustbot modify labels: T-doc, T-libs, C-enhancement, A-layout

**Comment by steveklabnik** [comment]
This is for the libs team to decide what the docs should say (I think the intent is that the "basically" implies things like your situation, but am not sure), and for them and the lang team to decide if what you're doing is legal.

**Comment by RalfJung** [comment]
> I think the intent is that the "basically" implies things like your situation

This is indeed the case. (IIRC, I wrote that sentence in the docs.)

**Comment by scottmcm** [comment]
For known ZSTs one can just use `mem::zeroed()` for less typing, no worries about weird `undef` stuff, and no perf hit.

Dunno if that's worth mentioning, though.  After all, `default()` is also free for ZSTs...

**Comment by RalfJung** [comment]
Note that not *all* ZSTs are in the exception of this "basically" -- clearly, `!` is not, and the same goes for enums with no variants and other uninhabited types.

**Comment by scottmcm** [comment]
Yes, absolutely.  I just meant that if it fell under the "basically" exception and it was a ZST, then one still doesn't really need to use `uninitialized`, and could use other, simpler things instead.

**Comment by steffahn** [comment]
I’ll link my use case here: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2d064fe8d7f579d0e59df9967861ee7a). It’s for closure types which don’t implement default (unfortunately). I make sure that the type is inhabited. Also usage of `zeroed` sounds weird on a zero-sized-type IMO, furthermore the documentation of `zeroed` doesn’t say anything about ZSTs either.

**Comment by jgarvin** [comment]
Can't `default()` for a ZST still be arbitrarily slow? Maybe not often but in principle I don't see why people can't define a complex `default` method for a ZST.

**Comment by danielhenrymantilla** [comment]
Uninit memory is only valid for padding bytes; so an inhabited type "where full padding bytes make for a valid representation of that type" are _valid_ to `mem::uninitialized()` / `MaybeUninit::uninit().assume_init()`. The inhabited types that uphold this property can be constructed out of the following three categories:

  - ZSTs (such as `()`, [_; 0]`)

  - unions where at least one variant is made of padding bytes (such as `MaybeUninit<_>`);

  - structural composition of such types (such as `[MaybeUninit<_>; …]`, `(MaybeUninit<_>,)`, _etc._).

Obviously if the type is uninhabited, then it can't be valid to produce instances of that type, and let's also mention, for completeness sake, that this is talking about _validity_ (well-defined to _produce_) and not about _safety_ _w.r.t._ library invariants (well-defined to feed to arbitrary library APIs).

  - For instance, even if [`GhostToken<'_>`](https://docs.rs/ghost-cell/0.2.2/ghost_cell/ghost_cell/struct.GhostToken.html) is an inhabited ZST, one can easily cause UB by forging an(y extra) instance of that type.

___

Back to the OP, you could also `transmute::<(), Zst>()`, if you are worried about the over-zealous uninit docs; of even `transmute_copy(&())` if in a generic context whereby Rust doesn't understand that the target type is a ZST.

**Comment by hkBst** [comment]
@steffahn what is the status of this?

And how does
`impl<T: Sized> AssertZeroSized for T {}`
not imply that all types that are `Sized` are zero-sized via:
```
trait AssertZeroSized: Sized {
    const ASSERT_ZERO_SIZED: () = [()][mem::size_of::<Self>()];
}
```
although I am not quite sure how to read `[()][mem::size_of::<Self>()]`.

BTW clippy now reports https://rust-lang.github.io/rust-clippy/master/index.html#uninit_assumed_init for line 33 (32 after rustfmt).

**Comment by RalfJung** [comment]
IMO this should be closed; we say "basically" because there are exceptions -- but they require very specific circumstances that I am not convinced we should spell out in the docs.

**Comment by steffahn** [comment]
I’m noticing that this section is new since the issue was opened:

> it is immediate undefined behavior to call this function on nearly all types, including integer types and arrays of integer types, and even if the result is unused.

I think that sufficiently clarifies that there are niche cases where it isn’t immediate undefined behavior. That all further details are left to the documentation of `MaybeUninit::assume_init` is completely appropriate for this deprecated function. The point of this issue was just about implications of the `mem::uninitialized` .

---

Re-reading these documentation today, I’m also noticing that “the function basically cannot be used correctly” can be interpreted in different ways: the term “used correctly” could refer not only to *any usage that is UB-free*, but also take into consideration the usage that matches the documented *purpose* of the API; and the *original intended* usage of this API is also relevant.

The documented purpose of this API is
> pretending to produce a value of type `T`, while doing nothing at all

and for almost all reasonable use-case that’s not what this API can do. You can’t *pretend* to produce a value of type `T` – if you call this function you'll always *actually* produce a value of type `T`. (Or hit UB.)

The [original](https://doc.rust-lang.org/1.30.0/std/mem/fn.uninitialized.html) intended usage of this API is that it’s the predecessor to `MaybeUninit`, setting out [specific usage rules](https://doc.rust-lang.org/1.30.0/std/mem/fn.uninitialized.html#undefined-behavior) which are no longer true, and also sketching some explicitly listed intended use-case in the introduction section as:
> This is useful for FFI functions and initializing arrays sometimes, but should generally be avoided.

both of which are generally instant-UB now in all nontrivial cases (i.e. for most types).

---

Having re-read this documentation now, I still have ideas for improvements (e.g. re-ordering some stuff), but  I can propose these separately; and all the relevant documented information appears to be there already.
